xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/objects/classes/DBListClass.java

```diff
@@ -282,8 +285,16 @@ public String getQuery(XWikiContext context)
                     select.append("doc." + idField);
                 } else {
                     select.append("idprop.value");
-                    fromStatements.add("StringProperty as idprop");
                     whereStatements.add("obj.id=idprop.id.id and idprop.id.name='" + idField + "'");
+                    // Get the from statements according to the type of property
+                    try {
+                        PropertyClass pc = (PropertyClass) context.getWiki().getDocument(classname, context).getXClass()
+                                .get(idField);
+                        String classType = (pc == null) ? "StringProperty" : pc.newProperty().getClass().getName();
+                        fromStatements.add(classType + " as idprop");
+                    } catch (XWikiException e) {
+                        e.printStackTrace();
```

Comment:
> Then "we consider that the default type is StringProperty" is not good. You should never explain what you do

Indeed, I agree with you, this should be intuitive for the reader , ( But  I got this information from the old behaviour of this function since the StringProperty is the most used ) , but I'll let it to NULL and hundel the exception.
I'll try to refactor the code and commit , 
Thanks Vincent